### PR TITLE
Generalize ProxyWatcher to monitor other resources

### PR DESCRIPTION
**Pull request by andrejtokarcik**
This makes the `ProxyWatcher` logic extensible and reusable for other resources like `Lock` (#7286).

### CODE REVIEW

**Review by Joerger — APPROVED** [review]

File: lib/services/watcher.go [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+		return nil, trace.Wrap(err)
+	}
+	return &ProxyWatcher{watcher, collector}, nil
+}
+
+// ProxyWatcher is built on top of resourceWatcher to monitor additions
+// and deletions to the set of proxies.
+type ProxyWatcher struct {
+	*resourceWatcher
+	*proxyCollector
```
Comment (smallinsky) [inline lib/services/watcher.go:271]: nit: I think that this is slightly overuse of struct embedding that impact readability. The consumer of ProxyWatcher doesn't use any resourceWatcher and proxyCollector methods or fields thus but embedding we are unnecessarily expose resourceWatcher and proxyCollector members and functions.

Comment (andrejtokarcik) [reply lib/services/watcher.go:271]: `ProxyWatcher` users call e.g. `Done` from `resourceWatcher` and `GetCurrent` from `proxyCollector`.

File: lib/services/watcher.go [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+		p.Log.Warningf("Unexpected event: %v.", event)
+		return nil
+	}
+
+	switch event.Type {
+	case types.OpDelete:
+		// Always broadcast when a proxy is deleted.
+		return p.updateCurrent(ctx, func() { delete(p.current, event.Resource.GetName()) }, true)
+	case types.OpPut:
+		_, known := p.current[event.Resource.GetName()]
```
[the remark this answers is not in the document]
Comment (andrejtokarcik) [reply lib/services/watcher.go:329]: Thanks for catching this!

File: lib/services/watcher.go [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+	*resourceWatcher
+	*proxyCollector
+}
+
+// proxyCollector accompanies resourceWatcher when monitoring proxies.
+type proxyCollector struct {
+	ProxyWatcherConfig
+	// current holds a map of the currently known proxies (keyed by server name,
+	// RWMutex protected).
+	current map[string]types.Server
```
[the remark this answers is not in the document]
Comment (andrejtokarcik) [reply lib/services/watcher.go:279]: I don't understand, sorry. The map keys are used to track the existence of resources (when processing events from the watcher) while the map values are aggregated in `GetCurrent` and when sending updates to `ProxiesC`.

[the remark this answers is not in the document]
Comment (smallinsky) [reply lib/services/watcher.go:279]: Right, I have missed that and this comment is not valid.

**Review by smallinsky — APPROVED** [review]

**Review by klizhentas — APPROVED** [review]

**Pull request merged**